File: examples/sphere_interpolation.py
#!/usr/bin/env python3
"""Discrete geodesic interpolation on the sphere across metric/retraction variants.

Python-API parallel of `examples/sphere_interpolation.cpp`. Runs
`discrete_geodesic` from the north pole to a shared target under five
configurations and prints per-case summary statistics:

  1. Round metric + exponential map (textbook great-circle).
  2. Round metric + projection retraction.
  3. Anisotropic metric A = diag(4, 1, 1).
  4. Anisotropic metric A = diag(25, 1, 1), smaller step size.
  5. Round metric near-antipodal target (graceful degradation near cut locus).

For each run we report status, iteration count, initial/final distance, path
length (waypoint count), arc length, and max angular deviation from the
great-circle plane through start and target. Anisotropic runs should show a
non-zero deviation — the path bends to trade expensive x-motion for cheaper
y-motion.
"""

import math

import numpy as np

import geodex


def spherical_point(theta: float, phi: float) -> np.ndarray:
    """Point on the unit sphere at polar angle theta (from north) and azimuth phi."""
    return np.array(
        [math.sin(theta) * math.cos(phi),
         math.sin(theta) * math.sin(phi),
         math.cos(theta)]
    )


def arc_length(points: list[np.ndarray]) -> float:
    """Sum of successive angular (great-circle) distances along the polyline."""
    total = 0.0
    for i in range(len(points) - 1):
        dot = np.clip(np.dot(points[i], points[i + 1]), -1.0, 1.0)
        total += math.acos(dot)
    return total


def max_plane_deviation(points: list[np.ndarray], start: np.ndarray,
                        target: np.ndarray) -> float:
    """Max angular deviation (radians) of any path point from the great circle
    plane spanned by `start` and `target`. Zero for a true great-circle arc."""
    normal = np.cross(start, target)
    n = np.linalg.norm(normal)
    if n < 1e-12:
        return 0.0  # degenerate (antipodal or identical endpoints)
    normal /= n
    return max(math.asin(min(1.0, abs(float(np.dot(p, normal))))) for p in points)


def run(label: str, manifold, start: np.ndarray, target: np.ndarray,
        settings: geodex.InterpolationSettings):
    result = geodex.discrete_geodesic(manifold, start, target, settings)
    pts = [np.asarray(p) for p in result.path]
    arc = arc_length(pts)
    dev = max_plane_deviation(pts, start, target)
    print(
        f"  [{label:48s}] status={str(result.status):24s} "
        f"iters={result.iterations:4d} "
        f"d_init={result.initial_distance:.4f} d_final={result.final_distance:.2e} "
        f"pts={len(pts):4d} arc={arc:.4f} rad  dev={math.degrees(dev):6.2f} deg"
    )


# Common start (north pole) and a shared target off the y=0 plane so anisotropy
# can bend the path along y.
north = np.array([0.0, 0.0, 1.0])
shared = spherical_point(theta=1.3, phi=0.9)  # ~52 deg azimuth

print("Running discrete_geodesic scenarios on the sphere:\n")

# 1. Round metric + exponential map (default Sphere).
s1 = geodex.InterpolationSettings(step_size=0.1)
run("1. Round metric, true exp/log", geodex.Sphere(), north, shared, s1)

# 2. Round metric + projection retraction.
s2 = geodex.InterpolationSettings(step_size=0.1)
run("2. Round metric, projection retraction",
    geodex.Sphere(retraction="projection"), north, shared, s2)

# 3. Moderate anisotropic metric via ConfigurationSpace wrapper.
A1 = np.diag([4.0, 1.0, 1.0])
aniso1 = geodex.ConfigurationSpace(geodex.Sphere(), geodex.ConstantSPDMetric(A1))
s3 = geodex.InterpolationSettings(step_size=0.1)
run("3. Anisotropic A=diag(4,1,1)", aniso1, north, shared, s3)

# 4. Heavy anisotropic metric — smaller step_size for stability.
A2 = np.diag([25.0, 1.0, 1.0])
aniso2 = geodex.ConfigurationSpace(geodex.Sphere(), geodex.ConstantSPDMetric(A2))
s4 = geodex.InterpolationSettings(step_size=0.05, max_steps=500)
run("4. Anisotropic A=diag(25,1,1)", aniso2, north, shared, s4)

# 5. Near-antipodal target (just inside the cut locus).
near_antipodal = spherical_point(theta=math.pi - 0.1, phi=0.9)
s5 = geodex.InterpolationSettings(step_size=0.2)
run("5. Round metric, near-antipodal", geodex.Sphere(), north, near_antipodal, s5)

File: examples/sphere_interpolation.cpp
/// @file sphere_interpolation.cpp
/// @brief Demonstrates `discrete_geodesic` on the sphere across several
/// metric / retraction combinations and writes a JSON file for visualisation
/// via `scripts/visualize.py`.
///
/// Run the example, then visualise:
/// ```
/// ./build/examples/sphere_interpolation sphere_interpolation.json
/// python scripts/visualize.py sphere_interpolation.json -o sphere.html
/// open sphere.html
/// ```

#include <cmath>
#include <cstdio>

#include <fstream>
#include <iomanip>
#include <iostream>
#include <numbers>
#include <string>
#include <vector>

#include <Eigen/Core>

#include "geodex/geodex.hpp"

using namespace geodex;

static Eigen::Vector3d point_at_theta(double theta) {
  return Eigen::Vector3d(std::sin(theta), 0.0, std::cos(theta));
}

/// Point on the unit sphere at polar angle `theta` from the north pole and
/// azimuth `phi` around the z axis (0 → x-axis, π/2 → y-axis).
static Eigen::Vector3d spherical_point(double theta, double phi) {
  return Eigen::Vector3d(std::sin(theta) * std::cos(phi), std::sin(theta) * std::sin(phi),
                         std::cos(theta));
}

/// Write a JSON array of 3D points: [[x,y,z], ...]
static void write_points(std::ofstream& out, const std::vector<Eigen::Vector3d>& pts) {
  out << "[";
  for (size_t i = 0; i < pts.size(); ++i) {
    if (i > 0) out << ",";
    out << "\n        [" << std::setprecision(8) << pts[i][0] << "," << pts[i][1] << ","
        << pts[i][2] << "]";
  }
  out << "\n      ]";
}

static void write_vec3(std::ofstream& out, const Eigen::Vector3d& v) {
  out << "[" << std::setprecision(8) << v[0] << "," << v[1] << "," << v[2] << "]";
}

/// A single path entry for the visualization JSON.
struct PathEntry {
  std::string label;
  Eigen::Vector3d start;
  Eigen::Vector3d target;
  std::vector<Eigen::Vector3d> points;
};

/// Run `discrete_geodesic` on the given sphere and return a labelled path entry
/// plus a short console summary.
template <typename SphereT>
static PathEntry run_scenario(const std::string& label, const SphereT& sphere,
                              const Eigen::Vector3d& start, const Eigen::Vector3d& target,
                              const InterpolationSettings& settings) {
  auto result = discrete_geodesic(sphere, start, target, settings);

  std::printf(
      "  [%-48s] status=%-17s iters=%4d halvings=%3d initial_d=%.4f final_d=%.2e points=%zu\n",
      label.c_str(), to_string(result.status), result.iterations, result.distortion_halvings,
      result.initial_distance, result.final_distance, result.path.size());

  return PathEntry{label, start, target, std::move(result.path)};
}

static void write_json(const std::string& output_file, const std::vector<PathEntry>& entries) {
  std::ofstream out(output_file);
  if (!out) {
    std::cerr << "Error: cannot open " << output_file << "\n";
    std::exit(1);
  }

  out << std::fixed;
  out << "{\n";
  out << "  \"manifold\": { \"type\": \"S2\" },\n";
  out << "  \"paths\": [\n";
  for (size_t i = 0; i < entries.size(); ++i) {
    const auto& e = entries[i];
    out << "    {\n";
    out << "      \"label\": \"" << e.label << "\",\n";
    out << "      \"start\": ";
    write_vec3(out, e.start);
    out << ",\n";
    out << "      \"target\": ";
    write_vec3(out, e.target);
    out << ",\n";
    out << "      \"points\": ";
    write_points(out, e.points);
    out << "\n    }" << (i + 1 == entries.size() ? "\n" : ",\n");
  }
  out << "  ]\n";
  out << "}\n";

  std::cout << "Wrote " << output_file << "\n";
}

int main(int argc, char* argv[]) {
  std::string output_file = "sphere_interpolation.json";
  if (argc > 1) output_file = argv[1];

  // Common start point at the north pole.
  const Eigen::Vector3d north(0.0, 0.0, 1.0);

  std::vector<PathEntry> entries;

  std::cout << "Running discrete_geodesic scenarios on the sphere:\n";

  // All anisotropic scenarios use a target with non-zero x AND y components so
  // the metric's directional preference is actually visible. A target on the
  // y=0 plane makes the problem one-dimensional along the great circle, and
  // any metric yields the same (straight great-circle) path.
  const double theta = 1.3;  // polar angle from north
  const double phi = 0.9;    // azimuth in x-y plane (~52 deg)
  const Eigen::Vector3d shared_target = spherical_point(theta, phi);

  // ---------------------------------------------------------------------
  // 1. Round metric + exponential map — the textbook great-circle geodesic.
  //    `log` is the Riemannian logarithm, `-log` is the natural gradient,
  //    and the path lies exactly on the great-circle arc.
  // ---------------------------------------------------------------------
  {
    Sphere<> sphere;  // default: SphereRoundMetric + SphereExponentialMap
    InterpolationSettings s;
    s.step_size = 0.1;
    entries.push_back(
        run_scenario("1. Round metric, true exp/log", sphere, north, shared_target, s));
  }

  // ---------------------------------------------------------------------
  // 2. Round metric + projection retraction — the retraction is first-order,
  //    but the metric is still the round one so the walk still follows the
  //    great circle. Provides a reference for "projection retraction alone"
  //    vs "anisotropic metric" effects.
  // ---------------------------------------------------------------------
  {
    Sphere<2, SphereRoundMetric, SphereProjectionRetraction> sphere;
    InterpolationSettings s;
    s.step_size = 0.1;
    entries.push_back(
        run_scenario("2. Round metric, projection retraction", sphere, north, shared_target, s));
  }

  // ---------------------------------------------------------------------
  // 3. Moderate anisotropic metric — A = diag(4, 1, 1) penalises motion in
  //    the x direction. The optimal path trades y-motion for less x-motion
  //    early on, producing a visible bend away from the great circle.
  //    `has_riemannian_log` is false for ConstantSPDMetric, so
  //    `discrete_geodesic` always uses the FD natural gradient here.
  // ---------------------------------------------------------------------
  {
    Eigen::Matrix3d A = Eigen::Matrix3d::Identity();
    A(0, 0) = 4.0;
    Sphere<2, ConstantSPDMetric<3>> sphere{ConstantSPDMetric<3>{A}};
    InterpolationSettings s;
    s.step_size = 0.1;
    entries.push_back(
        run_scenario("3. Anisotropic metric A=diag(4,1,1)", sphere, north, shared_target, s));
  }

  // ---------------------------------------------------------------------
  // 4. Heavy anisotropic metric — A = diag(25, 1, 1). The bend toward
  //    y-motion should be pronounced. Smaller step_size keeps the walk
  //    stable under the stretched metric.
  // ---------------------------------------------------------------------
  {
    Eigen::Matrix3d A = Eigen::Matrix3d::Identity();
    A(0, 0) = 25.0;
    Sphere<2, ConstantSPDMetric<3>> sphere{ConstantSPDMetric<3>{A}};
    InterpolationSettings s;
    s.step_size = 0.05;
    s.max_steps = 500;
    entries.push_back(
        run_scenario("4. Anisotropic metric A=diag(25,1,1)", sphere, north, shared_target, s));
  }

  // ---------------------------------------------------------------------
  // 5. Near-antipodal — shows graceful degradation near the cut locus. The
  //    log is still well-defined (distance < pi), so this should converge.
  // ---------------------------------------------------------------------
  {
    Sphere<> sphere;
    InterpolationSettings s;
    s.step_size = 0.2;
    const Eigen::Vector3d near_antipodal =
        spherical_point(std::numbers::pi - 0.1, 0.9);  // just inside cut
    entries.push_back(
        run_scenario("5. Round metric, near-antipodal", sphere, north, near_antipodal, s));
  }

  std::cout << "\n";
  write_json(output_file, entries);
  std::cout << "\nVisualise with:\n";
  std::cout << "  python scripts/visualize.py " << output_file << " -o sphere.html\n";
  return 0;
}
